umbral/openssl.py

```diff
@@ -1,218 +1,448 @@
-    or not on the curve.
+    Returns ``True`` if ``check_bn`` is in ``[0, modulus)``, ``False`` otherwise.
     """
     zero = backend._int_to_bn(0)
     zero = backend._ffi.gc(zero, backend._lib.BN_clear_free)
 
     check_sign = backend._lib.BN_cmp(check_bn, zero)
-    range_check = backend._lib.BN_cmp(check_bn, curve.order)
-    return check_sign == 1 and range_check == -1
+    range_check = backend._lib.BN_cmp(check_bn, modulus)
+    return (check_sign == 1 or check_sign == 0) and range_check == -1
 
 
-@typing.no_type_check
-def _int_to_bn(py_int: int, curve: 'Curve'=None, set_consttime_flag=True):
+def bn_from_int(py_int: int, check_modulus=None, set_consttime_flag=True):
     """
-    Converts the given Python int to an OpenSSL BIGNUM. If a curve is
-    provided, it will check if the Python integer is within the order of that
-    curve. If it's not within the order, it will raise a ValueError.
+    Converts the given Python int to an OpenSSL BIGNUM. If ``modulus`` is
+    provided, it will check if the Python integer is within ``[0, modulus)``.
 
     If set_consttime_flag is set to True, OpenSSL will use constant time
     operations when using this CurveBN.
     """
     conv_bn = backend._int_to_bn(py_int)
     conv_bn = backend._ffi.gc(conv_bn, backend._lib.BN_clear_free)
-    
-    if curve:
-        on_curve = _bn_is_on_curve(conv_bn, curve)
-        if not on_curve:
-            raise ValueError("The Python integer given is not on the provided curve.")
+
+    if check_modulus and not bn_is_normalized(conv_bn, check_modulus):
+        raise ValueError(f"The Python integer given ({py_int}) is not under the provided modulus.")
 
     if set_consttime_flag:
         backend._lib.BN_set_flags(conv_bn, backend._lib.BN_FLG_CONSTTIME)
     return conv_bn
 
-@typing.no_type_check
-def _bytes_to_bn(bytes_seq: bytes, set_consttime_flag=True):
+
+def bn_from_bytes(bytes_seq: bytes, set_consttime_flag=True, check_modulus=None, apply_modulus=None):
     """
-    Converts the given byte sequence to an OpenSSL BIGNUM. 
+    Converts the given byte sequence to an OpenSSL BIGNUM.
     If set_consttime_flag is set to True, OpenSSL will use constant time
     operations when using this BIGNUM.
     """
-    bn = _get_new_BN(set_consttime_flag)
+    bn = _bn_new(set_consttime_flag)
     backend._lib.BN_bin2bn(bytes_seq, len(bytes_seq), bn)
-    backend.openssl_assert(bn != backend._ffi.NULL)    
+    backend.openssl_assert(bn != backend._ffi.NULL)
+
+    if check_modulus and not bn_is_normalized(bn, check_modulus):
+        raise ValueError(f"The integer encoded with given bytes ({repr(bytes_seq)}) "
+                          "is not under the provided modulus.")
+
+    if apply_modulus:
+        bignum =_bn_new()
+        with backend._tmp_bn_ctx() as bn_ctx:
+            res = backend._lib.BN_mod(bignum, bn, apply_modulus, bn_ctx)
+            backend.openssl_assert(res == 1)
```

Comment:
It seems we're not doing anything with the `bignum` result later. It'd be cool that we could simply say `BN_mod(bn, bn, apply_modulus, bn_ctx)`, but it's not clear whether the API allows it.

Reply:
Good catch, we should indeed return `bignum` here.

I don't think any of the OpenSSL functions support inplace operations, unfortunately (seeing as how bignums probably use dynamic allocations inside).

Reply:
Fixed.